test/framework/src/main/java/org/elasticsearch/test/disruption/DisruptableMockTransport.java

```diff
@@ -0,0 +1,188 @@
+
+    protected abstract void handle(DiscoveryNode sender, DiscoveryNode destination, Runnable doDelivery);
+
+    private void sendFromTo(DiscoveryNode sender, DiscoveryNode destination, Runnable doDelivery) {
+        handle(sender, destination, new Runnable() {
+            @Override
+            public void run() {
+                if (getDisruptedCapturingTransport(destination).isPresent()) {
+                    doDelivery.run();
+                } else {
+                    logger.trace("unknown destination in {}", this);
+                }
+            }
+
+            @Override
+            public String toString() {
+                return doDelivery.toString() + " from " + sender + " to " + destination;
+            }
+        });
+    }
+
+    @Override
+    protected void onSendRequest(long requestId, String action, TransportRequest request, DiscoveryNode destination) {
+
+        assert destination.equals(getLocalNode()) == false : "non-local message from " + getLocalNode() + " to itself";
+        super.onSendRequest(requestId, action, request, destination);
+
+        final String requestDescription = new ParameterizedMessage("{}[{}] from {} to {}",
+            action, requestId, getLocalNode(), destination).getFormattedMessage();
+
+        final Runnable returnConnectException = new Runnable() {
+            @Override
+            public void run() {
+                handleError(requestId, new ConnectTransportException(destination, "disconnected"));
+            }
+
+            @Override
+            public String toString() {
+                return "disconnection response to " + requestDescription;
+            }
+        };
+
+        sendFromTo(getLocalNode(), destination, new Runnable() {
+            @Override
+            public void run() {
+                switch (getConnectionStatus(getLocalNode(), destination)) {
+                    case BLACK_HOLE:
+                        logger.trace("dropping {}", requestDescription);
+                        break;
+
+                    case DISCONNECTED:
+                        sendFromTo(destination, getLocalNode(), returnConnectException);
+                        break;
+
+                    case CONNECTED:
+                        Optional<DisruptableMockTransport> destinationTransport = getDisruptedCapturingTransport(destination);
+                        assert destinationTransport.isPresent();
+
+                        final RequestHandlerRegistry requestHandler = destinationTransport.get().getRequestHandler(action);
+
+                        final TransportChannel transportChannel = new TransportChannel() {
+                            @Override
+                            public String getProfileName() {
+                                return "default";
+                            }
+
+                            @Override
+                            public String getChannelType() {
+                                return "test-channel";
```

Comment:
Not sure where this string appears, but something more specific might be useful. Maybe `disruptable-mock-transport-channel`?